parlai/tasks/coco_caption/agents.py

```diff
@@ -0,0 +1,301 @@
+        return int(num_iters * self.num_episodes())
+
+    def _load_lens(self):
+        with open(self.length_datafile) as length:
+            lengths = json.load(length)
+            self.num_eps = lengths['num_eps']
+            self.num_exs = lengths['num_exs']
+
+    def _setup_data(self, test_info_path, annotation_path, unittest):
+        if not self.datatype.startswith('test'):
+            with open(annotation_path) as data_file:
+                self.annotation = json.load(data_file)
+        else:
+            with open(test_info_path) as data_file:
+                self.test_info = json.load(data_file)
+
+        if unittest:
+            if not self.datatype.startswith('test'):
+                self.annotation['annotations'] = self.annotation['annotations'][:10]
+            else:
+                self.test_info['images'] = self.test_info['images'][:10]
+
+        self.image_paths = set()
+        # Depending on whether we are using the train/val/test set, we need to
+        # find the image IDs in annotations or test image info
+        if not self.datatype.startswith('test'):
+            for anno in self.annotation['annotations']:
+                self.image_paths.add(self.image_path + '%012d.jpg' % (anno['image_id']))
+        else:
+            for info in self.test_info['images']:
+                self.image_paths.add(self.image_path + '%012d.jpg' % (info['id']))
+
+
+    def _setup_image_data(self):
+        '''hdf5 image dataset'''
+        extract_feats(self.opt)
+        im = self.opt.get('image_mode')
+        if self.opt.get('attention', False):
+            hdf5_path = self.image_path + 'mode_{}.hdf5'.format(im)
+        else:
+            hdf5_path = self.image_path + 'mode_{}_noatt.hdf5'.format(im)
+        hdf5_file = self.h5py.File(hdf5_path, 'r')
+        self.image_dataset = hdf5_file['images']
+
+        image_id_to_idx_path = self.image_path + 'mode_{}_id_to_idx.txt'.format(im)
+        with open(image_id_to_idx_path, 'r') as f:
+            self.image_id_to_idx = json.load(f)
+
+    def get_image(self, image_id):
+        if not self.use_hdf5:
+            im_path = self.image_path + '%012d.jpg' % (image_id)
+            return self.image_loader.load(im_path)
+        else:
+            img_idx = self.image_id_to_idx[str(image_id)]
+            return torch.Tensor(self.image_dataset[img_idx])
+
+    def num_examples(self):
+        if not self.datatype.startswith('test'):
+            return len(self.annotation['annotations'])
+        else:
+            return len(self.test_info['images'])
+
+    def num_episodes(self):
+        return self.num_examples()
+
+    def num_images(self):
+        if not hasattr(self, 'num_imgs'):
+            return self.num_examples()
+        return self.num_imgs
+
+
+class DefaultDataset(COCODataset):
+    pass
+
+
+class OeTeacher(FixedDialogTeacher):
+    """COCO Open-ended teacher
+    """
+    def __init__(self, opt, shared=None, version='2017'):
```

Comment:
We would like to have something similar to how open subtitles takes care of versions - that is, have two separate teachers for the 2014 and 2017 data. For example, if you specify on the command line ``coco_caption:v2017``, parlai will try to initialize a ``V2017Teacher`` which could be a subclass of this teacher.